discovery/gossiper.go

```diff
@@ -323,6 +323,112 @@ func (d *AuthenticatedGossiper) ProcessLocalAnnouncement(msg lnwire.Message,
 	return nMsg.err
 }
 
+// ChannelUpdateID is a unique identifier for ChannelUpdate messages, as
+// channel updates can be identified by the (ShortChannelID, Flags)
+// tuple.
+type ChannelUpdateID struct {
+	// channelID represents the set of data which is needed to
+	// retrieve all necessary data to validate the channel existence.
+	channelID lnwire.ShortChannelID
+
+	// Flags least-significant bit must be set to 0 if the creating node
+	// corresponds to the first node in the previously sent channel
+	// announcement and 1 otherwise.
+	Flags uint16
+}
+
+// deDupedAnnouncements de-duplicates announcements that have been
+// added to the batch. Internally, announcements are stored in three maps
+// (one each for channel announcements, channel updates, and node
+// announcements). These maps keep track of unique announcements and
+// ensure no announcements are duplicated.
+type deDupedAnnouncements struct {
+	// channelAnnouncements are identified by the short channel id field.
+	channelAnnouncements map[lnwire.ShortChannelID]lnwire.Message
+
+	// channelUpdates are identified by the channel update id field.
+	channelUpdates map[ChannelUpdateID]lnwire.Message
+
+	// nodeAnnouncements are identified by node id field.
+	nodeAnnouncements map[*btcec.PublicKey]lnwire.Message
+}
+
+// Reset operates on deDupedAnnouncements to reset storage of announcements
+func (d *deDupedAnnouncements) Reset() {
+	// Storage of each type of announcement (channel anouncements, channel
+	// updates, node announcements) is set to an empty map where the
+	// approprate key points to the corresponding lnwire.Message.
+	d.channelAnnouncements = make(map[lnwire.ShortChannelID]lnwire.Message)
+	d.channelUpdates = make(map[ChannelUpdateID]lnwire.Message)
+	d.nodeAnnouncements = make(map[*btcec.PublicKey]lnwire.Message)
+}
+
+// AddMsg adds a new message to the current batch.
+func (d *deDupedAnnouncements) AddMsg(message lnwire.Message) {
```

Comment:
This method should probably also not be exported, as the maps it accesses are not protected by mutexes, and it is only supposed to be used internally within the gossiper?

Reply:
Functionally, it doesn't matter if the method is exported or not. The type that the method targets itself is un-exported. Within the codebase we have a general convention of exporting methods of a private struct to indicate that those methods are meant to be used freely. 

A mutex isn't needed as this is only currently used within the main event loop of the gossiper, just as the prior slice used was never accessed concurrently.

Reply:
Ah, I did not know that! :D